R/schema_autobox.R

```diff
@@ -0,0 +1,216 @@
+#' especially the examples.
+#' @importFrom rlang !!!
+#'
+#' @return a `<config>` class object with list elements that can be arrays boxed.
+#' @export
+#'
+#' @examples
+#' config <- create_config(
+#'   create_rounds(
+#'     create_round(
+#'       round_id_from_variable = TRUE,
+#'       round_id = "origin_date",
+#'       model_tasks = create_model_tasks(
+#'         create_model_task(
+#'           task_ids = create_task_ids(
+#'             create_task_id("origin_date",
+#'               required = NULL,
+#'               optional = c(
+#'                 "2023-01-02",
+#'                 "2023-01-09",
+#'                 "2023-01-16"
+#'               )
+#'             ),
+#'             create_task_id("location",
+#'               required = "US",
+#'               optional = c("01", "02", "04", "05", "06")
+#'             ),
+#'             create_task_id("horizon",
+#'               required = 1L,
+#'               optional = 2:4
+#'             )
+#'           ),
+#'           output_type = create_output_type(
+#'             create_output_type_mean(
+#'               is_required = TRUE,
+#'               value_type = "double",
+#'               value_minimum = 0L
+#'             )
+#'           ),
+#'           target_metadata = create_target_metadata(
+#'             create_target_metadata_item(
+#'               target_id = "inc hosp",
+#'               target_name = "Weekly incident influenza hospitalizations",
+#'               target_units = "rate per 100,000 population",
+#'               target_keys = NULL,
+#'               target_type = "discrete",
+#'               is_step_ahead = TRUE,
+#'               time_unit = "week"
+#'             )
+#'           )
+#'         )
+#'       ),
+#'       submissions_due = list(
+#'         relative_to = "origin_date",
+#'         start = -4L,
+#'         end = 2L
+#'       )
+#'     )
+#'   )
+#' )
+#' schema_autobox(config)
+#' schema_autobox(config) |>
+#'   jsonlite::toJSON(
+#'     auto_unbox = TRUE, na = "string",
+#'     null = "null", pretty = TRUE
+#'   )
+schema_autobox <- function(config, box_extra_paths = NULL) {
+  checkmate::assert_list(box_extra_paths, null.ok = TRUE)
+  if (!inherits(config, "config")) {
+    cli::cli_abort(
+      c("x" = "{.arg config} must be an object of class {.cls config}
+             not {.cls {class(config)}}.")
+    )
+  }
+  schema_id <- attr(config, "schema_id")
+  schema <- download_tasks_schema(
+    hubUtils::extract_schema_version(schema_id)
+  )
+
+  # Get list of paths to config properties that can be arrays and may require
+  # boxing.
+  paths <- get_array_schema_paths(schema)
+  if (!is.null(box_extra_paths)) {
+    paths <- c(paths, box_extra_paths) |> unique()
+  }
+
+  # Identify and append any additional properties that may need to be boxed.
+  paths <- append_additional_properties(config, paths)
+
+  paths <- paths |>
+    purrr::map(~ expand_path_items(.x, config)) |>
```

Comment:
expand_path_items basically expands arrays of objects that in the schema are encoded as `"items"` to integer indices in an actual config file, according to the number of each element in the config.

Reply:
Not at all. Good idea.